File: test/proto/perl/smoke.pl
#!/usr/bin/env perl
# Smoke test for the Perl test provider port. Prints summary stats that must
# match the canonical TS output documented in PROVIDER.md.

use strict;
use warnings;
use FindBin;
use lib $FindBin::Bin;
require "$FindBin::Bin/Provider.pm";
Voxgig::Struct::Proto->import(qw(equal equal_strict error_matches struct_match));

sub main {
    my $prov = Voxgig::Struct::Proto->load;

    my $fns = $prov->functions;
    print 'functions: ', join(', ', @$fns), "\n";

    my $total = 0;
    my %expect_kinds;
    my %input_kinds;
    for my $fn (@$fns) {
        for my $entry (@{ $prov->entries($fn) }) {
            $total++;
            $expect_kinds{ $entry->{expect}{kind} }++;
            $input_kinds{ $entry->{input}{kind} }++;
        }
    }

    print "total entries: $total\n";

    # Fixed order to match the documented expected line.
    my @ek_order = qw(value absent match error);
    print 'expect kinds: ',
        join(', ', map { "$_=" . ($expect_kinds{$_} // 0) } @ek_order), "\n";

    my @ik_order = qw(in args ctx);
    print 'input kinds: ',
        join(', ', map { "$_=$input_kinds{$_}" } grep { $input_kinds{$_} } @ik_order),
        "\n";

    my $e = $prov->entries('getpath', 'basic')->[0];
    my $doc = $e->{doc} ? 'true(=1)' : 'false(=0)';
    printf
        "getpath/basic[0]: id=%s, doc=%s, input.kind=%s, expect.kind=%s, expect.value=%s\n",
        $e->{id}, $doc, $e->{input}{kind}, $e->{expect}{kind}, $e->{expect}{value};

    # ─── helper sanity checks ──────────────────────────────────────────────
    print 'equal(undef, missing) lenient: ', (equal(undef, undef) ? 'true' : 'false'), "\n";
    print 'equal_strict(undef, __NULL__): ',
        (equal_strict(undef, '__NULL__') ? 'true' : 'false'),
        ' / equal_strict(undef, 1): ',
        (equal_strict(undef, 1) ? 'true' : 'false'), "\n";
    print 'error_matches substring ci: ',
        (error_matches({ any => 0, text => 'Foo', regex => 0 }, 'a foobar error') ? 'true' : 'false'),
        "\n";
    my $sm = struct_match({ a => { b => 2 } }, { a => { b => 3 } });
    print 'struct_match failure: ',
        sprintf('{ok=%s, path=%s, expected=%s, actual=%s}',
            $sm->{ok} ? 1 : 0,
            join('/', @{ $sm->{path} // [] }),
            $sm->{expected} // '', $sm->{actual} // ''),
        "\n";
}

main();

File: test/proto/ocaml/smoke.ml
(* Smoke test for the OCaml test provider port. Prints summary stats that must
 * match the canonical TS output documented in PROVIDER work.
 *
 * NOTE: NOT compiled/run — the OCaml toolchain is unavailable in the authoring
 * environment. Self-reviewed against the canonical ts/provider.ts. *)

open Provider

(* Render an expect_kind / input_kind as its canonical lowercase tag. *)
let expect_kind_str = function
  | `Value -> "value"
  | `Error -> "error"
  | `Match -> "match"
  | `Absent -> "absent"

let input_kind_str = function `In -> "in" | `Args -> "args" | `Ctx -> "ctx"

(* Count occurrences, preserving a fixed display order. *)
let count_in_order order tbl =
  List.filter_map
    (fun k -> match List.assoc_opt k tbl with Some c when c > 0 -> Some (k, c) | _ -> None)
    order

let bump tbl k =
  let cur = match List.assoc_opt k !tbl with Some c -> c | None -> 0 in
  tbl := (k, cur + 1) :: List.remove_assoc k !tbl

let () =
  let prov = load () in

  let fns = functions prov in
  Printf.printf "functions: %s\n" (String.concat ", " fns);

  let total = ref 0 in
  let expect_kinds = ref [] in
  let input_kinds = ref [] in
  List.iter
    (fun fn ->
      List.iter
        (fun e ->
          incr total;
          bump expect_kinds (expect_kind_str e.expect.ekind);
          bump input_kinds (input_kind_str e.input.kind))
        (entries prov fn))
    fns;

  Printf.printf "total entries: %d\n" !total;

  (* Fixed order to match the documented expected line. *)
  let ek = count_in_order [ "value"; "absent"; "match"; "error" ] !expect_kinds in
  Printf.printf "expect kinds: %s\n"
    (String.concat ", " (List.map (fun (k, c) -> Printf.sprintf "%s=%d" k c) ek));

  let ik = count_in_order [ "in"; "args"; "ctx" ] !input_kinds in
  Printf.printf "input kinds: %s\n"
    (String.concat ", " (List.map (fun (k, c) -> Printf.sprintf "%s=%d" k c) ik));

  let e = List.hd (entries ~group:"basic" prov "getpath") in
  let id_str = match e.id with Some s -> s | None -> "null" in
  let value_str =
    match e.expect.value with Some v -> stringify v | None -> "null"
  in
  Printf.printf
    "getpath/basic[0]: id=%s, doc=%b, input.kind=%s, expect.kind=%s, expect.value=%s\n"
    id_str e.doc
    (input_kind_str e.input.kind)
    (expect_kind_str e.expect.ekind)
    value_str;

  (* ---- helper sanity checks ---- *)
  Printf.printf "equal(Null, Null) lenient: %b\n" (equal Null Null);
  Printf.printf "equal_strict(Null, __NULL__): %b / equal_strict(Null, 1): %b\n"
    (equal_strict Null (Str nullmark))
    (equal_strict Null (Num 1.0));
  Printf.printf "error_matches substring ci: %b\n"
    (error_matches { any = false; text = Some "Foo"; regex = false } "a foobar error");
  let sm =
    struct_match (Obj [ ("a", Obj [ ("b", Num 2.0) ]) ]) (Obj [ ("a", Obj [ ("b", Num 3.0) ]) ])
  in
  Printf.printf "struct_match failure: {ok=%b, path=%s, expected=%s, actual=%s}\n"
    sm.ok
    (String.concat "/" sm.path)
    (match sm.expected with Some v -> stringify v | None -> "")
    (match sm.actual with Some v -> stringify v | None -> "")
